gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/AsNumberStep.java

```diff
@@ -34,16 +34,24 @@
  */
 public class AsNumberStep<S> extends ScalarMapStep<S, Number> {
 
-    private N numberToken;
+    private GType typeToken;
 
     public AsNumberStep(final Traversal.Admin traversal) {
         super(traversal);
-        this.numberToken = null;
+        this.typeToken = null;
     }
 
-    public AsNumberStep(final Traversal.Admin traversal, final N numberToken) {
+    public AsNumberStep(final Traversal.Admin traversal, final GType numberToken) {
         super(traversal);
-        this.numberToken = numberToken;
+        if (!numberToken.isNumeric())
+            throw new IllegalArgumentException("asNumber() requires a numeric type token, got " + numberToken);
+        this.typeToken = numberToken;
+    }
+
+    // TODO potential extension to use class
+    public AsNumberStep(final Traversal.Admin traversal, final Class<?> numberToken) {
```

Comment:
Is the intention to keep this?

Nit: If it stays, the signature can be restricted slightly to

```suggestion
    public AsNumberStep(final Traversal.Admin traversal, final Class<? extends Number> numberToken) {
```